tests/task_scheduler_test.c

```diff
@@ -443,10 +444,64 @@ static int test_scheduler_oom_still_works(struct aws_allocator *allocator, void
     return AWS_OP_SUCCESS;
 }
 
-AWS_TEST_CASE(scheduler_pops_task_late_test, test_scheduler_pops_task_fashionably_late);
-AWS_TEST_CASE(scheduler_ordering_test, test_scheduler_ordering);
-AWS_TEST_CASE(scheduler_has_tasks_test, test_scheduler_has_tasks);
-AWS_TEST_CASE(scheduler_reentrant_safe, test_scheduler_reentrant_safe);
-AWS_TEST_CASE(scheduler_cleanup_cancellation, test_scheduler_cleanup_cancellation);
-AWS_TEST_CASE(scheduler_cleanup_reentrants, test_scheduler_cleanup_reentrants);
-AWS_TEST_CASE(scheduler_oom_still_works, test_scheduler_oom_still_works)
+static int s_test_scheduler_schedule_cancellation(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    tl_executed_tasks_n = 0;
+
+    struct aws_task_scheduler scheduler;
+    aws_task_scheduler_init(&scheduler, allocator);
+
+    struct aws_task task2;
+    aws_task_init(&task2, s_task_n_fn, (void *)2);
+
+    /* schedule 250 ns in the future. */
+    uint64_t task2_timestamp = 250;
+    aws_task_scheduler_schedule_future(&scheduler, &task2, task2_timestamp);
+
+    struct aws_task task1;
+    aws_task_init(&task1, s_task_n_fn, (void *)1);
+
+    /* schedule now. */
+    aws_task_scheduler_schedule_now(&scheduler, &task1);
+
+    struct aws_task task3;
+    aws_task_init(&task3, s_task_n_fn, (void *)3);
+
+    /* schedule 500 ns in the future. */
+    uint64_t task3_timestamp = 500;
+    aws_task_scheduler_schedule_future(&scheduler, &task3, task3_timestamp);
+
+    aws_task_scheduler_cancel_task(&scheduler, &task1);
+    aws_task_scheduler_cancel_task(&scheduler, &task2);
+
+    aws_task_scheduler_run_all(&scheduler, task3_timestamp);
```

Comment:
You try 1 more thing, cancelling a task that's just about to be run:
- schedule two more NOW tasks.
- the 1st one cancels the 2nd one